ocaml/typing/typecore.ml

```diff
@@ -2772,15 +2754,21 @@ and type_pat_aux
         pat_type = instance expected_ty;
         pat_attributes = sp.ppat_attributes;
         pat_env = !env }
-  | Ppat_constraint(sp_constrained, sty) ->
+  | Ppat_constraint(sp_constrained, sty, ms) ->
```

Comment:
Combined with another suggestion below, here we can assert `ms` to be empty.

This has the downside that, `let foo (x : a -> b -> c @@ modes)` will not work perfectly (the curry mode is not interpreted the best way), but it was like this before PR anyway.

Reply:
I don't think this is true -- `Ppat_constraints` with modes get created by, for example, `vb_pat_constraint`. The modes don't get removed before they reach this codepath.

Reply:
Yes you are right.